### ISSUE TITLE
Change key actions to be bitflags

### ISSUE THREAD

**Issue by Kerollmops** [meilisearch/meilisearch#4645]
Currently, the [Meilisearch action keys are single numbers](https://github.com/meilisearch/meilisearch/blob/abe29772dbebbfe5251cfeece2addd4133d3b93b/meilisearch-types/src/keys.rs) and therefore cannot be combined. It's not possible to OR them NOT them. I needed to do a basic `documents.add | documents.delete` for the edit documents by using a function feature.

We must [use bitflags to define those actions](https://lib.rs/crates/bitflags) and be able to combine them. We could have 32 or 64 different combinable actions. We already use this library in heed and it works very well.

**Comment by flevi29** [comment]
I would like to contribute. I'm pretty new to Rust, I'm not sure how to do this exactly.

I guess it should look something like this:
![image](https://github.com/user-attachments/assets/b35f5be5-8a98-436a-8b01-6623de853453)

But `Deserr` proc macro cannot be added
![image](https://github.com/user-attachments/assets/42a9df7f-9153-4174-9e51-fd4965d1f368)

And I'm not sure what's the problem here
![image](https://github.com/user-attachments/assets/ed7b387b-0aaf-4135-8242-b23527bbfc29)

I might be too new for this. Also this way would I have to assign the numbers incrementally manually for all the constants? So `... = 1; ... = 2; ... = 3; .....`?

**Comment by irevoire** [comment]
Hey, yes sorry it seems like the code generated by the `bitflags` crate is not compatible with deserr 😖 
You will need to implement `Deserr` by yourself.
Here's a few examples from the meilisearch repository itself:
- https://github.com/meilisearch/meilisearch/blob/9a756cf2c59a149e2eb83c1abd5932485c80c487/milli/src/vector/parsed_vectors.rs#L21-L50
- https://github.com/meilisearch/meilisearch/blob/9a756cf2c59a149e2eb83c1abd5932485c80c487/milli/src/vector/mod.rs#L329-L365
- And more if you search for `Deserr<E> for` in the codebase

I believe in your case it should look like this:
```rust
impl<E: DeserializeError> Deserr<E> for Action {
    fn deserialize_from_value<V: deserr::IntoValue>(
        value: deserr::Value<V>,
        location: deserr::ValuePointerRef<'_>,
    ) -> Result<Self, E> {
        match value {
            deserr::Value::String(s) => {
                match s {
                    "all" => Action::ALL,
                    "search" => Action::SEARCH,
                    ...
                }
            },
            value => Err(take_cf_content(E::error(
                None,
                deserr::ErrorKind::IncorrectValueKind {
                    actual: value,
                    accepted: &[deserr::ValueKind::String],
                },
                location,
            ))),
        }
    }
}
```

**Comment by flevi29** [comment]
`bitflags` wants me to do it something like this: https://docs.rs/bitflags/latest/bitflags/#custom-derives
Anyways, I'm going to return to this later, first I'll go through [`rustlings`](https://github.com/rust-lang/rustlings), as I am really unfamiliar with advanced rust code.

**Comment by flevi29** [comment]
I'm not sure how to use variant attributes:
![image](https://github.com/user-attachments/assets/09a74c0c-1236-499a-bf81-cba5c169187d)

**Comment by irevoire** [comment]
Hey @flevi29,

We don't use the `feature = "serde"` thing, if you directly derive `Serialize` or `Deserialize` does it works?

**Comment by flevi29** [comment]
You mean like this? Doesn't seem like it.
![image](https://github.com/user-attachments/assets/e76cc2ba-3a91-4c17-bc91-1d1df6bc4317)

**Comment by flevi29** [comment]
I could make a PR if that makes it easier to see what's going on with the faulty code.

**Comment by irevoire** [comment]
> You mean like this? Doesn't seem like it.

Yes, it’s supposed to be like this, but my guess is that it doesn't work with bitflags at all like deserr, and you'll also need to implement serialize and deserialize by hand 🥶

**Comment by flevi29** [comment]
@irevoire I've opened a draft PR. For now I'd just like to see what tests might fail for now, but my machine really struggles with `cargo test`. Can you please run the workflow?

**Comment by nnethercott** [comment]
yo @ManyTheFish is this still open or nah, seen a few PRs linked here but not sure if they closed it

**Comment by ManyTheFish** [comment]
Hey @nnethercott,

there is a tracking branch were I premerge the work, and flevi already finished the first part of the job.
There is a [PR created by another contributer](https://github.com/meilisearch/meilisearch/pull/5495) continuing the work but I am not sure he will finish it, however, it's a good starting implementation so if you want to tackle this issue, don't hesitate to start form this work.

The ultimate goal is being able to replace[ the following `actions::DOCUMENTS_ALL` by `documents.add | documents.delete`](https://github.com/meilisearch/meilisearch/blob/b43e895110332eecd4831fed3026a112980a520a/crates/meilisearch/src/routes/indexes/documents.rs#L1320)
